### PR TITLE
RHTAPINST-45: ArgoCD Declarative Token Creation

### PR DESCRIPTION
**Pull request by otaviof**
Replacing the `backing-services`'s `post-deploy.sh` hook with a Kubernetes native approach, using the same container images employed on other resources and ArgoCD itself.

See also https://github.com/argoproj/argo-cd/issues/9884

### CODE REVIEW

File: charts/rhtap-backing-services/templates/NOTES.txt [hunk 1]
```diff
@@ -1,13 +1,23 @@
...
 {{- end }}
+
+{{- if .Values.backingServices.argoCD.enabled }}
+  {{- $argoCD := .Values.backingServices.argoCD }}
+ArgoCD:
+  - Name: {{ $argoCD.name }}
+    Namespace: {{ $argoCD.namespace }}
+    Server URL: {{ include "backingServices.argoCD.serverHostname" . }}
+    Admin Secret: {{ include "backingServices.argoCD.secretClusterName" . }}
+    Integreation Secret: {{ $argoCD.secretName }}
```
Comment (Roming22) [inline charts/rhtap-backing-services/templates/NOTES.txt]: Integreation -> Integration

File: charts/rhtap-backing-services/templates/_copy-scripts.tpl [hunk 2]
```diff
@@ -0,0 +1,25 @@
+{{/*
+
+  POD container spec to copy scripts.
+
+*/}}
+{{- define "backingServices.copyScripts" -}}
+- name: copy-scripts
```
Comment (Roming22) [inline charts/rhtap-backing-services/templates/_copy-scripts.tpl:7]: Not related to this PR specifically, but I'm wondering if we should use init-containers in that pattern instead.

File: scripts/argocd-helper.sh [hunk 3]
```diff
@@ -0,0 +1,123 @@
+#!/usr/bin/env bash
+#
+# Tests if the ArgoCD instance is available on the cluster by logging in.
+#
+# Uses the ArgoCD session, created by previsoly running "argocd login", to
```
Comment (Roming22) [inline scripts/argocd-helper.sh]: previsoly -> previously

File: scripts/argocd-helper.sh [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+# generate a acccount token. The informaton is then stored on the ARGOCD_ENV_FILE
```
Comment (Roming22) [inline scripts/argocd-helper.sh]: a acccount -> an account
stored on -> stored in

File: scripts/argocd-helper.sh [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+# Retries a few times until the ArgoCD instance is available.
+test_argocd_login() {
+    for i in {1..30}; do
+        echo "# [${i}/30] Testing ArgoCD login on '${ARGOCD_HOSTNAME}'..."
+        argocd_login &&
+            return 0
+
+        wait=$((i * 5))
+        echo "### [${i}/30] Waiting for ${wait} seconds before retrying..."
```
Comment (Roming22) [inline scripts/argocd-helper.sh:57]: We should not sleep on the 30th loop as we won't try logging in again.

File: scripts/argocd-store-token.sh [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+# Waits for the environment file to be available.
+wait_for_env_file() {
+    echo "# Waiting for '${ARGOCD_ENV_FILE}' to be available..."
+    for i in {1..30}; do
+        [[ -r "${ARGOCD_ENV_FILE}" ]] &&
+            return 0
+
+        wait=$((i * 15))
+        echo "### [${i}/30] Waiting for '${ARGOCD_ENV_FILE}' to be available..."
+        sleep ${wait}
```
Comment (Roming22) [inline scripts/argocd-store-token.sh:33]: See previous comment about extra sleep.

File: scripts/argocd-store-token.sh [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+# Waits for the environment file to be available.
+wait_for_env_file() {
+    echo "# Waiting for '${ARGOCD_ENV_FILE}' to be available..."
+    for i in {1..30}; do
+        [[ -r "${ARGOCD_ENV_FILE}" ]] &&
+            return 0
+
+        wait=$((i * 15))
```
Comment (Roming22) [inline scripts/argocd-store-token.sh:31]: The step feels extra large. If my calculations are correct the max sleep time is 6975s, which seems overkill.

File: scripts/argocd-store-token.sh [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+        fail "ARGOCD_ENV_FILE is not set!"
+
+    wait_for_env_file ||
+        fail "ARGOCD_ENV_FILE='${ARGOCD_ENV_FILE}' not found or not readable!"
+
+    echo "# Sourcing ArgoCD credentials from '${ARGOCD_ENV_FILE}'..."
+    # shellcheck source=/dev/null
+    . "${ARGOCD_ENV_FILE}"
+
+    [[ -z "${ARGOCD_HOSTNAME}" ]] &&
```
Comment (Roming22) [inline scripts/argocd-store-token.sh:56]: Why not use `--from-file` and skip the validation?

File: charts/rhtap-backing-services/templates/_copy-scripts.tpl (hunk 2 above)
Comment (otaviof) [reply charts/rhtap-backing-services/templates/_copy-scripts.tpl:7]: That makes sense, +1

File: charts/rhtap-backing-services/templates/NOTES.txt (hunk 1 above)
Comment (otaviof) [reply charts/rhtap-backing-services/templates/NOTES.txt]: Just fixed the typo, thanks!

File: scripts/argocd-helper.sh (hunk 3 above)
Comment (otaviof) [reply scripts/argocd-helper.sh]: Thanks, fixed.

File: scripts/argocd-helper.sh (hunk 4 above)
Comment (otaviof) [reply scripts/argocd-helper.sh]: Also fixed 🙏

File: scripts/argocd-helper.sh (hunk 5 above)
Comment (otaviof) [reply scripts/argocd-helper.sh:57]: That's a fix we need to tackle on basically all scripts, good point. Could we tackle this in a single PR, please?

File: scripts/argocd-store-token.sh (hunk 7 above)
Comment (otaviof) [reply scripts/argocd-store-token.sh:31]: Before it happens the Helm Chart installation times out will kick in, but it's definitely extra long.

File: scripts/argocd-store-token.sh (hunk 8 above)
Comment (otaviof) [reply scripts/argocd-store-token.sh:56]: Interesting suggestion, I'll check that further.

File: scripts/argocd-store-token.sh (hunk 6 above)
Comment (otaviof) [reply scripts/argocd-store-token.sh:33]: Another fix/chore I'd like to tackle on a subsequent PR, please.

**Pull request merged**